lib/src/front_end/comment_writer.dart

```diff
@@ -0,0 +1,277 @@
+    // comments or it's needed to determine whether to insert a blank in a
+    // sequence.
+
+    // TODO(tall): If the token's comments are being moved by a fix, do not
+    // write them here.
+
+    var previousLine = _endLine(token.previous!);
+    var tokenLine = _startLine(token);
+
+    // Edge case: The analyzer includes the "\n" in the script tag's lexeme,
+    // which confuses some of these calculations. We don't want to allow a
+    // blank line between the script tag and a following comment anyway, so
+    // just override the script tag's line.
+    if (token.previous!.type == TokenType.SCRIPT_TAG) previousLine = tokenLine;
+
+    var comments = <SourceComment>[];
+    while (comment != null) {
+      var commentLine = _startLine(comment);
+
+      // Don't preserve newlines at the top of the file.
+      if (comment == token.precedingComments &&
+          token.previous!.type == TokenType.EOF) {
+        previousLine = commentLine;
+      }
+
+      var text = comment.lexeme.trim();
+      var linesBefore = commentLine - previousLine;
+      var flushLeft = _startColumn(comment) == 1;
+
+      if (text.startsWith('///') && !text.startsWith('////')) {
+        // Line doc comments are always indented even if they were flush left.
+        flushLeft = false;
+
+        // Always add a blank line (if possible) before a doc comment block.
+        if (comment == token.precedingComments) linesBefore = 2;
+      }
+
+      var type = CommentType.block;
+      if (text.startsWith('///') && !text.startsWith('////') ||
+          text.startsWith('/**') && text != '/**/') {
+        // TODO(rnystrom): Check that the comment isn't '/**/' because some of
+        // the dart_style tests use that to mean inline block comments. While
+        // refactoring the Chunk representation to move splits to the front of
+        // Chunk, I want to preserve the current test behavior. The fact that
+        // those tests pass with the old representation is a buggy quirk of the
+        // comment handling.
+        type = CommentType.doc;
+      } else if (comment.type == TokenType.SINGLE_LINE_COMMENT) {
+        type = CommentType.line;
+      } else if (commentLine == previousLine || commentLine == tokenLine) {
+        type = CommentType.inlineBlock;
+      }
+
+      var sourceComment =
+          SourceComment(text, type, linesBefore, flushLeft: flushLeft);
+
+      // TODO(tall): If this comment contains either of the selection endpoints,
+      // mark them in the comment.
+
+      comments.add(sourceComment);
+
+      previousLine = _endLine(comment);
+      comment = comment.next;
+    }
+
+    return (comments, tokenLine - previousLine);
+  }
+
+  /// Whether [comment] should be attached to the preceding token.
+  bool _isHangingComment(SourceComment comment) {
+    // Don't move a comment to a preceding line.
+    if (comment.linesBefore != 0) return false;
+
+    // Multi-line comments are always pushed to the next line.
```

Comment:
Doc and block comments can be a single line. `inlineBlock` comments can be multiple lines.

What is this "Multi-line" referring to? I don't think it can mean either of comments which _are_ or _can be_ multiple lines if the criteria is `doc || block`.

Reply:
Rewrote the comment to clarify. Also added some docs to the enum values.

The distinction between `CommentType.block` and `CommentType.inlineBlock` is a little weird. You get the latter for any C-style comment where there is also non-comment code on the same line as it. Whether that's a *good* distinction to make... I'm not sure. It's been like this for many years, so I'm leaving it alone for now.

This part of the code was basically lifted directly from the old formatter. (I didn't reuse it because eventually the old code will be deleted.)